### ISSUE TITLE
Adding a WAL checkpoint to dirty block upload thread can improve performance

### ISSUE THREAD

**Issue by hwertz** [s3ql/s3ql#324]
I found "rm -R"'ing a directory tree with a large number of files "ropes", I get nice fast deletes, a little delay, nice fast deletes, delay, etc..   I assume because the 4MB WAL fills and an insert/update then is "holding the bag" running an autocheckpoint, writing up to 4MB of changes from WAL to the database.   I do see code to implement "s3qlrm" does use smaller batches with a checkpoint built in, so really that's the way to go for removing a directory tree.  That said, I did patch WAL support into s3ql-4.0.0 at one point and found rm is faster with WAL than without even with the little pauses.  This patch doesn't fix that (and I don't think it's an issue, the delays were quite short anyway and overall wall time for rm'ing thousands of files is still good).  But it does eliminate delay under heavy I/O load (well as long as any writes are included.)

I did find I'd get a pretty hefty delay now and then when I had heavy disk I/O going to the cache, I assume because WAL autocheckpoint would kick in just wherever and stall the main s3ql loop until it finished.  Well (as long as there's write activity, not just reads and filesystem modifications) here's a nice solution to get the checkpoint off the main thread!

I placed a self.db.checkpoint() into _do_upload in block_cache.py.  This runs in the threads that runs asynchronously to upload blocks to the cloud or your local storage, so a WAL checkpoint happening there does not block the main thread and your filesystem keeps humming along!
  
I could make a patch file for  this, but it's a 1 line change. In database.py line 362, I did

```
             self.db.execute('UPDATE objects SET hash=NULL WHERE id=?', (obj_
+            self.db.checkpoint()
             await self.mlock.release(obj_id)
             await self.mlock.release(el.inode, el.blockno)
             async with self.transfer_completed:
```

The one downside appears to be a drop in terms of uploads/second if you have a large number of tiny files in the cache, I saw a drop (with 2 threads) from 70 a second to about 7 (which has like 180,000 files).  But, on the other hand, if you use normal sized files there seems to be no noticeable slowdown to the upload threads and s3ql is more responsive.

(Edit this seems to have mangled my code, but I added a self.db.checkpoint() just above where the await self.mlocks are in the code, and below the section where it either succeeds to upload a block or fails to and inserts a record into the DB accordingly.)

**Comment by hwertz** [comment]
OK never mind... I was testing with zstd-20 compression, with zlib-9 it's even worse, it still drops from 100s of writebacks a second to like 7-15 range, having a speed drop of 90% on writeback (even if just on small files) is definitely too high.

**Comment by Nikratio** [comment]
Thanks for looking into this! Could you provide some more details on how you measure the effect of this? (Ideally so that others can reproduce). 

Looking at https://www.sqlite.org/threadsafe.html, I think your patch should either have no effect (if your SQLite is in serialized mode, so database access by other threads will still block) or it will appear to work but eventually result in crashes and database corruption (if your SQLite is in multithreaded or single-threaded mode).

**Comment by hwertz** [comment]
As far as I can tell, zstd must have shockingly high startup overhead, like 1-2ms.  I don't know what it's doing, but if it's fed a lot of small files it seems to get nowhere near the MB/sec it does going through larger-sized blocks.  Anyway I did  test with zlib and it showed a large drop too.  Putting a checkpoint() there is not the way to go.  It still gets up to maybe 50 writes a second per thread, but these are on small files where it should be getting at least several hundred.

What I *did* find had a shocking speedup was just setting a larger WAL journal - I set PRAGMA wal_autocheckpoint=25000 (25x the default) in database.py.   This seems like it should be terrible, like you get all this speed then huge checkpoint delays that are like 25x as long, but it doesn't seem to extend the checkpoint time significantly, just give 25x the number of rm'ed or created files before you get a little fraction of a second delay.  I assume this is because data is not written randomly into the s3ql database but fairly localized, and the transactions in the WAL may be repeatedly updating the same blocks (in which case WAL checkpointing can consolidate these into a single write to the DB.) 

I tried 10000 first (40MB WAL) and that was already much faster than 1000 (4MB) , but 25000 (100MB) seemed even faster (longer time between writebacks without the time they took seeming to increase, and "seat of the pants" it didn't seem to be slowing down reads as the sqlite3 docs suggest a large WAL could.  I tried 65536 (256MB) but a) I think that WAS large enough to slow down reads a little and b) It just seemed excessive to run a 256MB WAL against a 180MB DB.

I did see on gkrellm with WAL autocheckpoint size of 1000 I would get a pretty constant 12MB/sec written to the database if i rm -R linux-6.4.2, but with a 10000 or 25000 I get maybe 500K a second writes to the WAL (with several thousand file a second deletion rate), then just the occasional burst of write traffic as the checkpiont runs.  Far less total disk I/O I think.  I suspect one big reason for this difference is in WAL mode checkpoints are the main point where sqlite3 actually fsync()s anything.

To test heavy activity I was just untar'ing linux-6.4.8.tar.xz, looking at s3qlstat to see what was waiting to be written back, testing out rm -Rv linux-6.4.8 (both when it was done writing back and when it wasn't) to see how that behaved.   As a stress test for heavy I/O I also have VirtualBox with some Ubuntu 18.04 and Ubuntu 22.04 disk image (that I can tell if disk I/O gets too slow because dmesg shows disk I/O timeouts.)  That seemed to work great with an increased journal too, even in the face of other I/O traffic to the filesystem.

The one potential downside, it does mean the .db-wal file is 100MB. Hopefully nobody is running a cache on a filesystem that tight on space that an extra 100MB file would be a big issue though.

**Comment by Nikratio** [comment]
I think it would help to untangle the various effects here. I'm still not sure what you measure when you say "speed-up", and I suspect it's different things in different contexts :-).

For starters, I would suggest to write a small benchmark problem that does nothing but compress a small  amount of data with various compression algorithms. That will tell us what the compression overhead is, regardless of database properties.

Then let's quantify what we mean by "speed" of database operations. For example, we could time how long it takes to execute `s3qlcp` or `s3qlrm` on a fixed directory tree, depending on WAL journal size and checkpointing interval. This should be independent of compression algorithm, since nothing is compressed. It seems to me that  this is what your first message refers to.

In your last message, "speed-up" seems instead to refer to the speed of extracting the linux kernel. Is that right? Or are you timing the duration of "rm -r" of the extracted directory? If so, do you wait until the cache is clean? Or do you actively flush it? It would be great to have a sequence of operations that I can execute myself and see the same results.

**Comment by hwertz** [comment]
I did both, I waited for the cache to clear out on some runs, and ran some rm -Rf while the cache was still running on others.  I was just watching the tar and rm -Rv run and see how fast the text was scrolling by really, and saw how long the "hitches" are as it flushes WAL.  I didn't time it with the "time" command but it was visibly faster.

I realized what was happening with the I/O I measured was simply heavy deletes or writes causing multiple MB/sec of writes into WAL (I think this was not reliably showing up in gkrellm because the writes are cached, sqlite3 checkpoints definitely runs an fsync() but I'm not sure when fsync is used when writing into the WAL log).  I could probably use iostat or something on this to measure more reliably rather than just watching a little bar graph creep along in gkrellm.  And at least run tar and rm under "time" so I get some wall clock times.  

Based on a run I did with wal_autocheckpoint=0 (not recommended obviously, the WAL then grows without bounds -- well, unless you run s3qlcp or s3qlrm which have a checkpoint in the implementations), based on the speed the WAL grew in tests I suspect the 4MB default WAL flushes multiple times a second under heavy file creation/deletion loads.   Conversely, I went and did some runs with "find" and running "file *" to fire off some read activities and the WAL only went up by a matter of KB (since I imagine it's only updating atime.)

I do agree, running (probably a partial checkpoint, where it does the least blocking of read/write activities) could be good.  Since all DB activity goes through the code in database.py, perhaps there's some way to track actual number of write transactions (insert, update, etc.) and have a thread do a checkpoint when the count gets high enough.  

That said, I was pretty shocked at the speeds I saw giving it a larger WAL though, I don't have a s3ql-4.0.0 install around to compare to (I do have it here so I could reinstall for testing if needed...) but I think s3ql-5 is easily outrunning it now on filesystem metadata-heavy operations, and I found s3ql-4.0.0 to be plenty fast enough at it.  On regular operations it's about the same (since after all, grabbing a block from storage, decompressing it and putting it into the cache; or I/O on cached data, it's not database-heavy and dominated by I/O and decompression time... mostly I/O since zlib and zstd both decompress at 100s of MB/sec.).

**Comment by Nikratio** [comment]
My apologies, but I'm getting a little lost know. Could you summarize, in one sentence, what this issue is about and how I can tell if it has been fixed?

**Comment by hwertz** [comment]
I ran some timing tests this evening/night:
I put a line in database.py to execute "PRAGMA wal_autocheckpoint = xxxx" and varied the value xxxx.  I had been using 25000 (4096 byte page) WAL size but decided to test some values.  The 30 second dirty block upload delay undoubtedly increased the times on ones with smallest WAL size (and longest run times) but nevertheless the difference in time based on WAL size is striking.:

(dirty block upload delay is set to 30 seconds)
(l.tar just contains linux-6.4.8/Documentation)
time tar -xf ~/l.tar ; time rm -R linux-6.4.8

1000 -- 32 seconds create 32 delete
1024 -- 31 seconds create 31 seconds delete
2048 -- 20 seconds create 26 seconds delete
4096 -- 16 seconds create 15 delete
8192 -- 14 seconds create 12 delete
16384 -- 13 seconds create 9 delete
32768 -- 12 seconds create 8 seconds delete
65536 -- 12 seconds create 7 seconds delete
----------
larger data set -- Documentation and arch
time tar -Jxvf ~/l2.tar ; time rm -R linux-6.4.8
4096 -- 60 seconds create 54 seconds delete
8192 -- 44 seconds create 40 delete
16384 -- 43 seconds create 36 seconds delete
32768 -- 38 seconds create 36 seconds delete
--------------

It really looks like the "sweet spot" here is 8192 pages (32MB), or possibly somewhere between 4096 and 8192 -- times drop off rapidly up to that point, with a much smaller decrease as you crank the WAL autocheckpoint size significantly higher.  Which per sqlite docs could start slowing down DB read speeds (I didn't test this), and of course potentially could mean a large checkpoint delay if transactions got in the WAL that were slow to write back to DB although I think s3ql's relatively "non-random" reads/writes to the DB make this less likely than it could be for some sqlite databases with different use patterns.

**Comment by hwertz** [comment]
Sure.  This was just investingating speedups to s3ql through adjusting the WAL (write ahead log) parameters, or possibly doing WAL checkpoints off the main thread.

**Comment by Nikratio** [comment]
> Sure. This was just investingating speedups to s3ql through adjusting the WAL (write ahead log) parameters, or possibly doing WAL checkpoints off the main thread. Enhancement rather than a bug.

Ok. Can you clarify the test that you propose to use to determine what the optimal parameters are?

I think that for e.g. tarball extraction benchmark we will always get the best speed by making the WAL just large enough so that it can contain all the changes from that operation (i.e., no checkpoint is needed). The load is also write-heavy, so slowdowns from reading a large WAL should not matter much. Unfortunately this does not mean that the corresponding WAL size is the right one in all situations.

**Comment by hwertz** [comment]
I thought that might be the case; but in fact if I set autocheckpoint=0, the larger "l2" file creation test finished in 28 seconds, deletion in 25, WAL grows at an alarming rate though, 877M at the end of the file creation test and 1.4GB after I delete them (so somewhere between 20-30MB/sec?)    I wonder if sqlite WAL uses a minimum of 1 page per transaction?  There's a total of 27590 files and directories in there but 877M WAL still seems like a lot for that (I mean the entire DB for that 1TB drive is 216MB). Anyway, the test sets are definitely getting plenty of checkpoints in.

It's true though, making sure a 8192 page WAL doesn't hose reads would be good.  I haven't noticed any speed dropoffs here but have not run a timed test yet on that.  I suppose I would probably turn off checkpoints (since otherwise I don't know if the WAL has just been checkpointed and is essentially empty with autocheckpoints on...) and do enough activity to get a roughly 4MB, 16MB, 64MB, 128MB, and 256MB WAL and run some timed "find", "ls -R" and maybe "file *" in some directories where I already have the files cached (I use keep-cache so that part's easy enough),  (I did find that read-intensive activity does not increase WAL very quickly at all since I suppose it's just updating some atimes or the like.)  I have to assume as the WAL gets up to 100s of MB it has to have some effect but iI'll run some timed measurements and see.  (Not tonight but soonish.)

**Comment by hwertz** [comment]
To summarize, I think the performance increase I saw increasing WAL checkpoint from default 1000 pages (about 4MB) to 8192 (32MB) was due to the surprisinngly heavy WAL activity during mass file creation/deletion, when I tested with autocheckpoints off I saw WAL increase by 20-30MB/sec.  This would make the WAL checkpoint (and relatively slow fsync to the database) run 5-10 times a second on default settings; going to 32MB WAL reduced it to about every 1.5-2 seconds.   Also why increasing WAL further had a relatively minor impact, the big speedup is from avoiding near-constant fsyncs.

I can run some read testing to see what impact a larger WAL has on metadata read speeds (i.e. time walking through some directory trees and such; I can try reading some already-cached files too, although I assume that time will be dominated by the actual file reads.)  I do have one disk hoked to a "slowputer" with a Celeron N3050 (Atom-style CPU), running "find" on a linux source tree only takes 6 seconds on one of my regular systems so it'll probably take 30 seconds on that POS, easier to discern a small like 1-5% speed difference if it's there.

**Comment by hwertz** [comment]
I ran a quick test -- (not on slowputer) -- with checpoint off, I untar'ed or rm'ed some files to get WAL size up then ran find on a directory tree with 399367 files and directories in it (some source code) with "time find darling > /dev/null".  I ran each at least 5 times, unfortunately I was seeing rather high variation in times (plus or minus about 1/4 second) but several runs would have a consistent value and averaging all runs also ended up with that value.

near-0 WAL (672K), time was 16.8s.  With 283MB WAL, time was 17.2s.  670MB WAL, 17.6 seconds.  This shows a 4.8% slowdown with 670MB WAL, 2.4% with 283MB (and, 670MB is 2.36x the size of 283, and I got 2x the time... which is probably measurement erorr but shows the measurements are at least roughly correct.)  Extrapolating from these values (assuming time goes up linearly with size), a 32MB value should yield about a 0.2-0.3% slowdown when WAL is totally full compared to a completely empty WAL (compared to 1/8th that value, roughly 0.02-0.04% for 4MB). Given the measurement jitter I was seeing I doubt I could distinguish between a 32MB WAL, 4MB, and empty WAL.

**Comment by Nikratio** [comment]
Sorry for letting this linger for so long! Do I understand correctly that you're now suggesting to increase the WAL checkpoint from 1000 pages to a higher value to get better performance instead of doing explicit checkpointing in a separate thread? Or to do both? And what is the value that you recommend based on your testing?

**Comment by hwertz** [comment]
I found having the checkpoint run in a seperate thread was not helpful.  At least where I put it, the checkpoint was firing off far too often and I found this did not help performance.  I think the root cause here is the checkpointing has relatively high overhead, and checkpoints were occuring very frequently under untar'ing or "rm -R" type activities.  Perhaps having a timed thread that checkpointed at a regular time interval (like every second or two) would be useful, then the WAL size could just adapt to the level of file system activity.  I don't know.

I *did* find setting WAL to 8192 is helpful though.  

When I set PRAGMA wal_autocheckpoint=0 (for testing, WAL grows without bounds so I could monitor how many MB/sec are going into it...), I found the tested activities were generating a solid 20-30MB/sec of WAL.  So default autocheckpoint size is checkpointing many times a second, a good 8 times a second or so.  Presumably the overhead of that is high enough that doing less frequent, larger checkpoints gets a nice speedup.

I found putting in PRAGMA wal_autocheckpoint=8192 got me a nice speedup, while past that there are rapidly diminishing returns.   Going from default 1000 to 4096 cut times right in half, while 8192 cut the untar time by about 56% and the "rm -R" time by about 62%.   Or time-wise, it shaves 32 second operation down by 18 seconds for the tar and 20 for the file deletions, while going arbitrarily large only shaved a further 2 seconds off the untar test and 5 off the deletions.  (The second test, which was actually larger than the first, shows even less speedup past WAL size of 8192, there's diminishing returns there.)

Finally, I tested to make sure a larger WAL did not decrease speed of walking the filesystem, I attempted to measure the difference in time between a 32MB and 4MB WAL.  The difference was too small for me to measure.  So I tested a few much larger WAL sizes, found the slowdown is linear with the size of the WAL, and extrapolated the slowdown from a 32MB WAL is about 0.2%.

**Comment by Nikratio** [comment]
Thanks for the summary! So it seems to be setting `PRAGMA wal_autocheckpoint=8192` and closing the issue with that is the best course of action here?

**Comment by hwertz** [comment]
Yes I would recommend that.
Side note, I do intend to get back on finishing up the zstd patches.  I have it running here and it's pretty nice (zstd 20-22 are dead slow, but I'm running zstd-19 to get nice compression with an OK speed, zstd-15 is already a hair faster than zlib-9 is.)
Thanks!
-Henry
